modules/axtask/src/run_queue.rs

```diff
@@ -84,49 +257,56 @@ impl AxRunQueue {
     pub fn exit_current(&mut self, exit_code: i32) -> ! {
         let curr = crate::current();
         debug!("task exit: {}, exit_code={}", curr.id_name(), exit_code);
-        assert!(curr.is_running());
+        assert!(curr.is_running(), "task is not running: {:?}", curr.state());
         assert!(!curr.is_idle());
         if curr.is_init() {
-            EXITED_TASKS.lock().clear();
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.clear());
             axhal::misc::terminate();
         } else {
             curr.set_state(TaskState::Exited);
-            curr.notify_exit(exit_code, self);
-            EXITED_TASKS.lock().push_back(curr.clone());
-            WAIT_FOR_EXIT.notify_one_locked(false, self);
-            self.resched(false);
+
+            // Notify the joiner task.
+            curr.notify_exit(exit_code);
+
+            // Push current task to the `EXITED_TASKS` list, which will be consumed by the GC task.
+            EXITED_TASKS.with_current(|exited_tasks| exited_tasks.push_back(curr.clone()));
+            // Wake up the GC task to drop the exited tasks.
+            WAIT_FOR_EXIT.with_current(|wq| wq.notify_one(false));
+            // Schedule to next task.
+            self.inner.resched(false);
         }
         unreachable!("task exited!");
     }
 
-    pub fn block_current<F>(&mut self, wait_queue_push: F)
-    where
-        F: FnOnce(AxTaskRef),
-    {
+    pub fn blocked_resched(&mut self) {
         let curr = crate::current();
-        debug!("task block: {}", curr.id_name());
-        assert!(curr.is_running());
-        assert!(!curr.is_idle());
-
-        // we must not block current task with preemption disabled.
-        #[cfg(feature = "preempt")]
-        assert!(curr.can_preempt(1));
+        assert!(curr.is_blocked(), "task is not blocked, {:?}", curr.state());
 
-        curr.set_state(TaskState::Blocked);
-        wait_queue_push(curr.clone());
-        self.resched(false);
+        debug!("task block: {}", curr.id_name());
+        self.inner.resched(false);
     }
 
+    /// Unblock one task by inserting it into the run queue.
+    /// If task's `on_cpu` flag is true,
+    /// it will enter a loop until the task finishes its scheduling process,
+    /// see `unblock_locked` for details.
     pub fn unblock_task(&mut self, task: AxTaskRef, resched: bool) {
-        debug!("task unblock: {}", task.id_name());
-        if task.is_blocked() {
+        task.clone().unblock_locked(|| {
```

Comment:
Can we put `fn unblock_locked` here?

Reply:
Yep, but `unblock_lock` is a field of axtask structure, I do not want to expose it to other mods